### PR TITLE
[Speculative decoding 1/9] Optimized rejection sampler

### PR DESCRIPTION
**Pull request by cadedaniel**
### Speculative decoding
This PR is a part of a larger series of PRs implementing speculative decoding, contributed to open source vLLM by Anyscale. See https://github.com/vllm-project/vllm/pull/2188 and [Speculative decoding open sourcing plan](https://docs.google.com/document/d/1rE4pr3IdspRw97XbImY4fS9IWYuJJ3HGtL7AdIKGrw8/edit#heading=h.qijw1sdidrer) for more information.

### Rejection sampling
This PR implements optimized rejection sampling, including the following features:
* Implementation of modified rejection sampling as described in https://arxiv.org/pdf/2302.01318.pdf
* All operations are batched on GPU, allowing non-blocking computation.
* Efficient collection of metrics regarding acceptance rate and number of emitted tokens.

It also contributes tests which verify the rejection sampler's ability to approximate distributions, given enough samples.

The following people contributed to it: @cadedaniel @Yard1 @amogkam

### Details

The basic idea behind [rejection sampling](https://en.wikipedia.org/wiki/Rejection_sampling) is that one can sample from the target distribution (larger model) using samples from a proposal distribution (smaller draft model), while guaranteeing the output distribution is equivalent to the target distribution.

"Modified" rejection sampling is introduced in the paper. It ensures that at least one token will always be emitted from the rejection sampling routine, even if all proposal tokens are rejected.

With LLMs, modified rejection sampling can reduce latency because multiple proposal sequences can be evaluated at once (batching on the GPU).

Finally, the paper introduces the notion of a "bonus" token. In the case where all proposed tokens are accepted, an additional token can be emitted. This is possible by having the target model predict the next token given the entire proposed sequence as context.

Visual confirmation that modified rejection sampling approximates the target distribution:
rejsample1
rejsample2

code for visualizations: https://gist.github.com/cadedaniel/07c1cd4ac003f51140b205580ac02613

### CODE REVIEW

**Comment by cadedaniel** [comment]
cc @LiuXiaoxuanPKU @WoosukKwon @zhuohan123 @Yard1

File: vllm/model_executor/layers/rejection_sampler.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+        .. math::
+            x_{n+1} \sim (q(x|x_1, \dots, x_n) - p(x|x_1, \dots, x_n))_+
+
+        where :math:`(f(x))_+` is defined as:
+
+        .. math::
+            (f(x))_+ = \frac{\max(0, f(x))}{\sum_x \max(0, f(x))}
+
+        See https://github.com/anyscale/vllm/pull/157 for a visualization of the
```
Comment (pcmoritz) [inline vllm/model_executor/layers/rejection_sampler.py:217]: Should be updated to https://github.com/vllm-project/vllm/pull/2336 :)

**Comment by cadedaniel** [comment]
The next PR will be https://github.com/cadedaniel/vllm-public/pull/1, will create it once this is merged.

File: vllm/model_executor/layers/rejection_sampler.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        bonus_token_ids = bonus_token_ids.squeeze()
+        batch_size, k = recovered_token_ids.shape
+
+        # Determine the index of the first False value for each row.
+        limits = (accepted == 0).max(1).indices
+        limits[~(accepted == 0).any(1)] = k
+
+        # Create masks using the indices.
+        indices = torch.arange(k, device=accepted.device).unsqueeze(0)
+        accepted_mask = indices < limits.unsqueeze(1)
```
Comment (LiuXiaoxuanPKU) [inline vllm/model_executor/layers/rejection_sampler.py:279]: what's the difference between `accepted` and `accepted_mask`?

File: vllm/model_executor/layers/rejection_sampler.py [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+
+        Args:
+            strict_mode: Whether or not to perform shape/device/dtype checks
+                during sampling. This catches correctness issues but adds
+                nontrivial latency.
+        """
+        super().__init__()
+        self.probs_dtype = torch.float32
+        self.token_id_dtype = torch.int64
+        self._num_bonus_tokens = 1
```
Comment (LiuXiaoxuanPKU) [inline vllm/model_executor/layers/rejection_sampler.py:26]: when can `num_bonus_tokens` > 1? Is it the last generated token by the target model iff all drafted tokens are accepted?

Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py]: > when can num_bonus_tokens > 1?

It is always 1. This variable is for readability only. I'll add a comment.

> Is it the last generated token by the target model iff all drafted tokens are accepted?

Yep!

File: vllm/model_executor/layers/rejection_sampler.py (hunk 2 above)
Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py:283]: `accepted` is the result of the rejection sampling condition. `accepted_mask` is `True` up until the first position rejected by the rejection sampling condition.

Example for k=3, bs=5:
```python
>>> accepted
tensor([[ True, False,  True],
        [False, False, False],
        [ True,  True, False],
        [ True,  True,  True],
        [False,  True, False]])
>>> accepted_mask
tensor([[ True, False, False],
        [False, False, False],
        [ True,  True, False],
        [ True,  True,  True],
        [False, False, False]])
```

**Comment by zhaoyang-star** [comment]
Very exciting work! I hope this feature can be merged soon as many other framwork such as TGI, TRT-LLM, llama.cpp, gpt-fast have supported Speculative sampling.

File: vllm/model_executor/layers/rejection_sampler.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        difference = target_probs - draft_probs
+
+        # TODO(cade): Can we use logprobs instead of probs, and avoid the
+        # division-by-zero errors without introducing distribution drift?
+
+        # shape [batch_size, k, vocab_size]
+        f = torch.clamp(difference, min=self._smallest_positive_value)
+
+        # shape [batch_size, k, vocab_size]
+        recovered_probs = f / torch.sum(f, dim=-1).reshape(-1, k, 1)
```
Comment (LiuXiaoxuanPKU) [inline vllm/model_executor/layers/rejection_sampler.py:243]: Nit: `torch.multinomial` does not require the probability to be normalized.

File: .github/workflows/scripts/test_scripts/run_spec_decode_tests.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
Comment (LiuXiaoxuanPKU) [inline .github/workflows/scripts/test_scripts/run_spec_decode_tests.sh:1]: We can remove this file for now because we have not set up the CI

File: vllm/model_executor/layers/rejection_sampler.py (hunk 4 above)
Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py:243]: I will leave this in to keep the maths consistent with https://arxiv.org/pdf/2302.01318.pdf. This operation is not the compute or scheduling bottleneck.

File: .github/workflows/scripts/test_scripts/run_spec_decode_tests.sh (hunk 5 above)
Comment (cadedaniel) [reply .github/workflows/scripts/test_scripts/run_spec_decode_tests.sh]: removed

**Review by LiuXiaoxuanPKU — APPROVED** [review]
